### PR TITLE
PROD-22203 Accommodate scheduler bounces without downtime.

### PR DESCRIPTION
**Pull request by bennorton767**

### CODE REVIEW

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -361,13 +362,87 @@ def run(args, dag=None):
...
+            latest_heartbeat = session.query(LJ).filter(LJ.id == ti.job_id).first().latest_heartbeat
+            if latest_heartbeat >= limit_dttm:
+
+                real_ti = session.query(TI).filter(
+                    TI.dag_id == ti.dag_id,
+                    TI.task_id == ti.task_id,
+                    TI.execution_date == ti.execution_date
+                ).with_for_update().first()
+
+                if real_ti.state == State.RUNNING:
```
Comment (bennorton767) [inline airflow/bin/cli.py]: This is a razor-thin race condition since we return if we're in state running up above.  However, it is possible for the state to be set to running between our initial check, and our check where we acquire a lock.

File: airflow/jobs.py [hunk 2]
```diff
@@ -1432,34 +1432,6 @@ def _process_executor_events(self, simple_dag_bag, session=None):
...
-                        dag = dagbag.get_dag(dag_id)
-                        ti.task = dag.get_task(task_id)
-                        ti.handle_failure(msg)
-                    except Exception:
-                        self.log.error("Cannot load the dag bag to handle failure for %s"
-                                       ". Setting task to FAILED without callbacks or "
-                                       "retries. Do you have enough resources?", ti)
-                        ti.state = State.FAILED
-                        session.merge(ti)
-                        session.commit()
```
Comment (bennorton767) [inline airflow/jobs.py]: This seems like a scary thing to do - ripping out a whole "feature" in airflow.  Allow me to explain myself:

*CHANGE SUMMARY:*

This block was added in the following PR:
https://github[dot]com/apache/airflow/pull/2715
which was written to address the following ticket:
https://issues.apache[dot]org/jira/browse/AIRFLOW-1641

(The above URLS are de-fanged to avoid programatic posting of this comment directly onto upstream's repo or jira.  While I believe everything I'm saying here needs to be documented for our records, I don't feel we gain anything by shouting our displeasure through a megaphone.  In the future our airflow repo will be private and we will not need to take this kind of measure.)

This code lives inside of SchedulerJob._process_executor_events() - this method's job is to retrieve the exit status of celery executors on the workers, and report on that exit status.

This block I am removing has the following effect:
"If the celery executor finished, but the state of the task instance is still pending, fail the task instance."

I need to remove this block because it is now intended for celery executors to exit without changing the state of the task instance if the hostname is set from a previous worker instance, and the heartbeat has not expired.

*CHANGE JUSTIFICATION:*

The following is a reasonably compact summary of the timeline of events limited to the pieces that were relevant to my decision to drop this feature rather than tweak its implementation:

- The author of the ticket (The Reporter) asserted that his tasks were getting stuck in a pending state because of a timeout loading the dag in `DagBag.process_file()` . This code runs on the worker prior to setting state to running.  If a timeout occurs, the worker swallows the error and the celery task returns successfully, but the task instance state isn't changed from pending, and the task is not run.
- The author of the pull request (The Implementer) put up a PR and requested that The Reporter test his changes because he (The Implementer) hadn't tested his own changes.  (I mention this because at no point in the ticket or the pull request does anybody assert that they have tested that this feature behaves as intended, and there are mistakes here that I believe would have been caught if someone had tested it, like the visibly incorrect log statement on 1447-1448.)
- The first two comments on this block of code in the PR from airflow maintainers are "Is there nothing better we can do here?" and "I share this concern."
- After addressing code read comments, The Implementer asserts that he is uncomfortable with the state of these changes because they introduce an inconsistency between the way BackfillJobs are handled when compared to other tasks.

In addition to the lack of confidence the above timeline inspires, I also feel strongly that this implementation is incorrect.  If a timeout importing dags occurs, the celery executor should fail, and that should trigger a task failure.

You'll notice reading this that I didn't go ahead and change this procedure to what I describe above as the correct implementation.  There are several reasons for this:

- The dag importing code is used many places in airflow, so removing the swallowing of dag import timeouts is a surprisingly invasive change.
- We do not have the problem of dags failing to import within the timeout, and I hesitate to invest much engineering time into fixing edge cases we do not have that only arise when airflow is abused.  (Dag generation should not be expensive.)
- Doing this correctly involves making a variety of changes to airflow that would be difficult to merge into our upcoming upstream patch.

After consideration of all of the above, I elected to simply remove this "feature" entirely from airflow.  I do not think we will miss it, and if we do fall prey to the issue "fixed" by this "feature", it should be a quick matter of inspecting the logs to see that this is what has happened.

*TL;DR:*
This code is a problem, so I'm removing it.  I think it's unlikely we'll regret this, and it's easy to reverse if we do.

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -361,13 +362,87 @@ def run(args, dag=None):
 
     task = dag.get_task(task_id=args.task_id)
     ti = TaskInstance(task, args.execution_date)
-    ti.refresh_from_db()
-    if ti.current_state() == "failed":
-        # This behavior is always correct (failed tasks should not run), but is implemented to specifically handle the
-        # case where a worker goes down, the celery task gets re-queued, and the scheduler kills the airflow task
-        # without knowing to revoke the task from the celery queue.
+    ti.refresh_from_db(include_job_id=True)
```
Comment (bennorton767) [inline airflow/bin/cli.py]: This is the beginning of the "new worker" side of the scheduler bounce recovery scenario.  Most of the new logic is here.  This is in `cli.py - run()` which is what airflow workers do when they are given a task instance to execute.

File: airflow/jobs.py [hunk 4]
```diff
@@ -2619,14 +2591,14 @@ def on_kill(self):
 
     @provide_session
     def heartbeat_callback(self, session=None):
-        """Self destruct task if state has been moved away from running externally"""
+        """Reset state to running if state set to null, scheduled, or queued externally.  Self-destruct if state is set to any other state externally."""
```
Comment (bennorton767) [inline airflow/jobs.py]: This is the beginning of the "old worker" side of the scheduler bounce recovery scenario.  This is in `LocalTaskJob.heartbeat_callback()` which is responsible for responding to external mutations to the task's state.

File: airflow/models.py [hunk 5]
```diff
@@ -1092,7 +1152,7 @@ def error(self, session=None):
         session.commit()
 
     @provide_session
-    def refresh_from_db(self, session=None, lock_for_update=False, include_queue_time=False):
+    def refresh_from_db(self, session=None, lock_for_update=False, include_queue_time=False, include_job_id=False):
```
Comment (bennorton767) [inline airflow/models.py]: I'm continuing in our storied tradition of only including new fields out of refresh_from_db if they're explicitly called for.  I instigated this pattern when I discovered (to my horror) that simply always pulling all the fields from the database when asked to update a TaskInstance will break parts of airflow.  Frankly I'm afraid to investigate why, and we don't really have time, so we have some patented _cowardly params_ ™️  to help us out.

File: airflow/bin/cli.py [hunk 6]
```diff
@@ -361,13 +362,87 @@ def run(args, dag=None):
 
     task = dag.get_task(task_id=args.task_id)
     ti = TaskInstance(task, args.execution_date)
-    ti.refresh_from_db()
-    if ti.current_state() == "failed":
-        # This behavior is always correct (failed tasks should not run), but is implemented to specifically handle the
```
Comment (dangermike) [inline airflow/bin/cli.py]: deleting this comment is really funny

File: airflow/bin/cli.py [hunk 7]
```diff
@@ -361,13 +362,87 @@ def run(args, dag=None):
...
-        # case where a worker goes down, the celery task gets re-queued, and the scheduler kills the airflow task
-        # without knowing to revoke the task from the celery queue.
+    ti.refresh_from_db(include_job_id=True)
+    from airflow.utils.state import State
```
Comment (dangermike) [inline airflow/bin/cli.py]: I'm afraid to ask -- why does this import have to be local?

Comment (dangermike) [reply airflow/bin/cli.py]: If this can be a module-level import, we can also have a module-level set for non-retryable states

Comment (bennorton767) [reply airflow/bin/cli.py]: I'm pretty sure it doesn't - I put it there out of fear.  I will move it up to the top level and see if it breaks anything.

File: airflow/bin/cli.py (hunk 6 above)
Comment (bennorton767) [reply airflow/bin/cli.py]: It is funny :)

In the interest of being overly clear, this comment remains true - this check is just now a subset of the conditions insisted upon in 367-374

File: airflow/jobs.py (hunk 2 above)
Comment (dangermike) [reply airflow/jobs.py]: you are doing the Lord's work

File: airflow/models.py [hunk 8]
```diff
@@ -371,6 +376,61 @@ def kill_zombies(self, session=None):
...
+                    # in-memory map.  This would naturally lead one to the conclusion that something like the below was
+                    # needed here:
+                    #
+                    # if ti_key in self.executor.queued_tasks:
+                    #     del self.executor.queued_tasks[ti_key]
+                    # if ti_key in self.executor.running:
+                    #     del self.executor.running[ti_key]
+                    #
+                    # What you would find, however, is that the above does not work.  The reason it doesn't work is
+                    # a matter of some regrettable (some might say unconscionable) complexity:
```
Comment (dangermike) [inline airflow/models.py]: how do you really feel?

File: airflow/models.py [hunk 9]
```diff
@@ -1594,22 +1656,28 @@ def signal_handler(signum, frame):
                 Stats.gauge(
                     'ti_status',
                     value=0,
-                    tags=[
+                tags=[
```
Comment (dangermike) [inline airflow/models.py]: is this tabbed correctly?

File: airflow/models.py [hunk 10]
```diff
@@ -1594,22 +1656,28 @@ def signal_handler(signum, frame):
...
                         'task_id:%s' % self.task_id,
                         'dag_id:%s' % self.dag_id,
                         'operator:%s' % self.task.__class__.__name__,
                     ],
                 )
             self.refresh_from_db(lock_for_update=True)
+            self.hostname = None
```
Comment (dangermike) [inline airflow/models.py]: Is this to ensure that any tasks pulled think they are on a new host and are released?

File: airflow/models.py (hunk 9 above)
Comment (bennorton767) [reply airflow/models.py]: Whoops!  It is not.  Fixed.  (batching changes up for testing - will be in the next commit)

File: airflow/models.py (hunk 10 above)
Comment (bennorton767) [reply airflow/models.py]: This is at the end of `TaskInstance._run_raw_task()`, so the objective here is to unset host before the task completes successfully.

In this happy-path case where the task succeeded, this doesn't really do anything - if this hostname unset goes through, the task state change to succeeded will go through, which will prevent the scheduler bounce recovery scenario from occurring.  Still, I think it's nice to be able to query the DB for task instances that have hostname set and be able to say that all of the results are still pending completion.

File: airflow/jobs.py (hunk 2 above)
Comment (alexagriffith) [reply airflow/jobs.py]: hmm... im blocked from those links

File: airflow/models.py [hunk 11]
```diff
@@ -371,6 +376,61 @@ def kill_zombies(self, session=None):
...
+                    #
+                    # Don't worry, it does get more confusing.  The skeptical programmer may add logging to this method
+                    # (kill_zombies) to inspect the state of the executor, and see that updates to the executor state
+                    # in the primary SchedulerJob find their way into the executor of the secondary SchedulerJobs.
+                    # Well, that's because the subprocess that ends up calling kill_zombies is repeatedly generated, and
+                    # evey time it is created, it inherits the executor from the parent process, with all of the
+                    # updated state that entails.
+                    #
+                    # While it is laudable to separate the work of generating dags from the work of scheduling
+                    # those dags, one could reasonably ask WHY THESE WERE NOT MADE SEPARATE THINGS GIVEN THAT THEY HAVE
```
Comment (alexagriffith) [inline airflow/models.py]: hahaha

File: airflow/models.py [hunk 12]
```diff
@@ -371,6 +376,61 @@ def kill_zombies(self, session=None):
...
+                    # DIFFERENT JOBS AND RUN IN DIFFERENT PROCESSES.
+                    #
+                    # Even having accepted the unknowability of the above, once could still find themselves wondering
+                    # WHY THE SCHEDULING WORK OF IDENTIFYING AND LABELING ZOMBIES IS PERFORMED BY THE PROCESS WHOSE
+                    # RESPONSIBILITY IS THE GENERATION OF DAGS.
+                    #
+                    # Unfortunately, the way this zombie killing responsibility is "designed", the only way to fix
+                    # the problem of the "real SchedulerJob" not knowing that the "dag processing SchedulerJob"
+                    # identified a zombie is to speak through some inter-process channel, which seems a bit like
+                    # buying a jet because your family dentist lives in another country.
```
Comment (alexagriffith) [inline airflow/models.py]: this is great

File: airflow/jobs.py (hunk 2 above)
Comment (bennorton767) [reply airflow/jobs.py]: Those links are de-fanged - you'll need to replace [dot] with .
